### PR TITLE
Separate distro tests to be per distribution

### PR DESCRIPTION
**Pull request by rjernst**
The java based distribution tests currently have a single Tests class
which encapsulates all of the tests for a particular distribution. The
test task in gradle then depends on all distributions being built, and
each individual tests class looks for the particular distribution it is
trying to test. This means that reproducing a single test failure
triggers all the distributions to be built, even though only one is
needed for the test.

This commit reworks the java distribution tests to pass in a particular
distribution to be tested, and changes the base test classes to be
actual test classes which have assumptions around which distributions
they operate on. For example, the archives tests will be skipped when
run with an rpm distribution, and vice versa for the package tests. This
makes reproduction much more granular. It also also better splitting up
tests around a particular use case. For example, all tests for systemd
behavior can be in one test class, and run independently of all tests
against rpm/deb distributions.

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/vagrant/VagrantShellTask.java [hunk 1]
```diff
@@ -104,15 +104,15 @@ public void runScript() {
                     spec.setProgressHandler(progressHandler);
                 });
             } catch (Exception e) {
-                getLogger().error("Failed command, dumping dmesg", e);
+                /*getLogger().error("Failed command, dumping dmesg", e);
```
Comment (atorok) [inline buildSrc/src/main/java/org/elasticsearch/gradle/vagrant/VagrantShellTask.java]: Leftover ?  Some more commented code above

**Review by alpar-t — APPROVED** [review]
LGTM, left some minor comments

File: buildSrc/src/main/java/org/elasticsearch/gradle/LazyString.java [hunk 2]
```diff
@@ -17,15 +17,25 @@
...
 
-package org.elasticsearch.packaging.test;
+package org.elasticsearch.gradle;
 
-import org.elasticsearch.packaging.util.Distribution;
+import java.util.function.Supplier;
 
-public class DefaultNoJdkDebBasicTests extends PackageTestCase {
+/**
+ * An functional interface to allow a String supplier to be resolved when calling toString().
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/LazyString.java]: nit: this isn't really a functional interface in the pedantic sense of `@FunctionalInterface`.

File: buildSrc/src/main/java/org/elasticsearch/gradle/LazyString.java [hunk 3]
```diff
@@ -17,15 +17,25 @@
...
+ *
+ * This can be used in many places where a string expression should not be evaluated until runtime.
+ */
+public class LazyString {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/LazyString.java]: This class doesn't seem to be used anywhere.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/GradleDistroTestTask.java [hunk 4]
```diff
@@ -79,7 +79,7 @@ public void extraArg(String arg) {
         line.append(taskName);
         line.append(" --project-cache-dir ");
         line.append(isWindows ? convertWindowsPath(getProject(), cacheDir) : convertLinuxPath(getProject(), cacheDir));
-        line.append(" -S");
+        //line.append(" -S");
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/GradleDistroTestTask.java]: Can this be removed?

File: buildSrc/src/main/java/org/elasticsearch/gradle/vagrant/VagrantMachine.java [hunk 5]
```diff
@@ -75,7 +75,7 @@ public void execute(Action<VagrantExecSpec> action) {
             execSpec.setEnvironment(System.getenv()); // pass through env
             execSpec.environment("VAGRANT_CWD", vagrantfile.getParentFile().toString());
             execSpec.environment("VAGRANT_VAGRANTFILE", vagrantfile.getName());
-            execSpec.environment("VAGRANT_LOG", "debug");
+            //execSpec.environment("VAGRANT_LOG", "debug");
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/vagrant/VagrantMachine.java]: Can we ditch this too?

File: qa/os/windows-2012r2/build.gradle [hunk 6]
```diff
@@ -1,11 +1,12 @@
+import org.elasticsearch.gradle.test.GradleDistroTestTask
 
 String boxId = project.properties.get('vagrant.windows-2012r2.id')
 if (boxId != null) {
   vagrant {
     hostEnv 'VAGRANT_WINDOWS_2012R2_BOX', boxId
   }
 } else {
-  tasks.named('distroTest').configure {
+  tasks.withType(GradleDistroTestTask).configureEach {
```
Comment (mark-vieira) [inline qa/os/windows-2012r2/build.gradle]: We should probably include a comment here as to "why" we are doing this. Even if it's temporary.

**Review by mark-vieira — APPROVED** [review]
Minor comments otherwise LGTM. This looks like it will make folks working on packaging tests (ie... you) very happy 😄

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java [hunk 7]
```diff
@@ -109,17 +112,26 @@ public void apply(Project project) {
...
+                }
+            }
+
+
             batsTests.forEach((desc, task) -> {
                 configureVMWrapperTask(vmProject, desc, task.getName(), vmDependencies).configure(t -> {
                     t.setProgressHandler(new BatsProgressLogger(project.getLogger()));
-                    t.onlyIf(spec -> vagrant.isWindowsVM() == false); // bats doesn't run on windows
+                    t.onlyIf(spec -> isWindows(vmProject) == false); // bats doesn't run on windows
+                    t.dependsOn(copyDistributionsTask);
```
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java]: Yes, unfortunately.

File: buildSrc/src/main/java/org/elasticsearch/gradle/LazyString.java (hunk 3 above)
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/LazyString.java]: Removed. This was leftover from an avenue I explored.

File: buildSrc/src/main/java/org/elasticsearch/gradle/vagrant/VagrantMachine.java (hunk 5 above)
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/vagrant/VagrantMachine.java]: I uncommented these for now. I will remove them in a separate, focused commit, when we are sure they are no longer useful.

File: qa/os/windows-2012r2/build.gradle (hunk 6 above)
Comment (rjernst) [reply qa/os/windows-2012r2/build.gradle]: This isn't temporary, it's just disable the distribution tests for windows when we don't have a windows image to test with. I've added a comment.

**Comment by rjernst** [comment]
@mark-vieira I believe I addressed all your feedback.

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/packaging

**Comment by mark-vieira** [comment]
Changes LGTM.

**Pull request merged**